README.md
# lib_BL999

Arduino library for **BL999** remote Temperature/Humidity sensor<br/>
It supports both **BL999** sensor types: with humidity and temperature-only

## Table of Contents

  1. [Sensor description](#sensor-description)
  1. [Protocol description](#protocol-description)
  1. [Structure of data bits in a message](#structure-of-data-bits-in-a-message)
  1. [Bits purpose](#bits-purpose)
  1. [Library usage](#library-usage)
  1. [API reference](#api-reference)

## Sensor description
  * BL999 sensor working frequency is 433 Mhz
  * Max distance: 30 m
  * Temperature range -40 +50 Celsius
  * Humidity 1-99%

## Protocol description
  * BL999 uses PWM to send it's data.
  * BL999 message packet consists of `one` start bit and `36` value bits divided by PWM high pulses
  * BL999 repeats same message `4` times during one data sending session. 
  * Usual PWM oscillogram looks like:
  
```
  ___         ___     ___  ___
  | |         | |     | |  | |
  | |         | |     | |  | |
__| |_________| |_____| |__| |_ 

   |     |         |      |
    
   H1    L1        L2     L3
```
  
  * H1: PWM high pulse which is used as a divider of data bits. 
    It's length is `~550 micro seconds ± 100 microseconds`.
    All high pulses should looks like H1 in BL999 message packet.
  * L1: PWM low pulse which is used as a start bit marker. 
    It's length is `~9 millis ± 1 millis`
    It should always be present at the beginning of the message packet
  * L2: Data bit which represents binary `1`. It's length is `~3.9 ms ± 0.5 ms`
  * L3: Data bit which represents binary `0`. It's length is `~1.85 ms ± 0.5 ms`
  * Whole message consists of 36 data bits: 
    binary `1` or `0` (which looks like L2 or L3) which are divided by H1 high pulses
  
  
## Structure of data bits in a message

  * data bits in a message are grouped by nibbles (4 bit groups)
  * thus there are totally 9 nibbles in a message
 
```
 nibble:        T1    ||     T2    ||     T3    ||     T4    ||     T5
 bit:      A0|A1|A2|A3||B0|B1|B2|B3||C0|C1|C2|C3||D0|D1|D2|D3||E0|E1|E2|E3

 nibble:        T6    ||     T7    ||     T8    ||     T9
 bit:      F0|F1|F2|F3||G0|G1|G2|G3||H0|H1|H2|H3||I0|I1|I2|I3
```

## Bits purpose
  * **A0-A3,B2-B3**  - UUID which is randomly set once sensor is turned on
                and it is not changed until next power off

  * **B0-B1**  - Number of channel of the sensor: 
                 `01` - 1, `10` - 2, `11` - 3

  * **C0** - battery info: `0` - battery is ok, `1` - battery is low
  
  * **C1-C3** - unknown

  * **D0-F3** - temperature written backwards and multiplied by 10. For example:
         `1010|0011|0000` = 19.7° (Celsius) because `0000|1100|0101` = 197
         (note that not only nibbles written in reversed order, but bits in nibble also reversed)
         negative temperature is stored in [two's complement code](https://en.wikipedia.org/wiki/Two%27s_complement), 
         for example:
         `0010|1001|1111` = -10.8° (Celsius) because
         `01101100` (inverted all bits and added 1) = 108

  * **G0-H3** - humidity written backwards as `100 - HUMIDITY` in 
          [two's complement code](https://en.wikipedia.org/wiki/Two%27s_complement), 
          for example:
          `1110|1011` = 59% because `101001` (inverted all bits and add 1) = 41, and 100 - 41 = 59

  * **I0-I3** - 4 check sum which is calculated as a sum of less significant bits
            of check sum of the nibbles T1-T8 written backwards.
            for example let's look at the example message:
            `1011|1100|0100|0000|1111|0000|0010|1111|0010`
            sum of the nibbles T1-T8 where each bit written backwards = 110100,
            lets take last 4 bits and wrote them backward:
            0010 which is equal to T9 nibble
            
## Library usage

  * This library is intended to be used with 433 Mhz receiver which is able to work in 
    [OOK](https://en.wikipedia.org/wiki/On-off_keying) mode.
    For example like [this](http://randomnerdtutorials.com/rf-433mhz-transmitter-receiver-module-with-arduino/) one.
    Please read datasheet for your receiver about how to use it with Arduino and how to use it in OOK mode
  * In the example below signal pin of the receiver 
    is expected to be attached to Arduino digital pin 2.
    Please note that only digital pins which supports ISR should be used.
    See [Digital Pins suitable for ISR on Arduino](https://www.arduino.cc/en/Reference/AttachInterrupt).
    Sketch infinetely wait for message from any BL999 transmitter
    and writes it to serial port
  
```C
#include <lib_bl999.h>

static BL999Info info;

void setup() {
    Serial.begin(115200);

    //set digital pin to read info from
    bl999_set_rx_pin(2);

    //start reading data from sensor
    bl999_rx_start();
}

void loop() {

    //waits infinitely until message will be received.
    //Either correct or not.
    //check sum does not checked at this moment
    bl999_wait_rx();

    //read message to info and if check sum correct - outputs it to the serial port
    if (bl999_get_message(info)) {
        Serial.println("====== Got message from BL999 sensor: ");
        Serial.println("Channel: " + String(info.channel));
        Serial.println("PowerUUID: " + String(info.powerUUID));
        Serial.println("Battery is " + String(info.battery == 0 ? "Ok" : "Low"));
        Serial.println("Temperature: " + String(info.temperature / 10.0));
        Serial.println("Humidity: " + String(info.humidity) + "%");
    }
}
```

  * The next example is the same except that in this example main loop does not blocked at all
  
```C
#include <lib_bl999.h>

static BL999Info info;

void setup() {
    Serial.begin(115200);

    //set digital pin to read info from
    bl999_set_rx_pin(2);

    //start reading data from sensor
    bl999_rx_start();
}

void loop() {

    //read message to info and if check sum correct - outputs it to the serial port
    //does not block main loop
    if (bl999_have_message() && bl999_get_message(info)) {
        Serial.println("====== Got message from BL999 sensor: ");
        Serial.println("Channel: " + String(info.channel));
        Serial.println("PowerUUID: " + String(info.powerUUID));
        Serial.println("Battery is " + String(info.battery == 0 ? "Ok" : "Low"));
        Serial.println("Temperature: " + String(info.temperature / 10.0));
        Serial.println("Humidity: " + String(info.humidity) + "%");
    }
}
```  



## API reference


<table>
    <thead>
        <tr>
            <th>Function/Structure</th>
            <th>Description</th>
            <th>Details</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>BL999Info</td>
            <td>stucture which holds sensor information like channel, temperature, humidity and power UUID</td>
            <td>
                <pre>
typedef struct {
   // up to 3 channels
   byte channel : 2;
   
   //unique power state per 
   //current power on of the sensor
   byte powerUUID : 6;
   
   // 0 - ok, 1- low
   byte battery : 1;
   
   // temperature * 10, 
   // for example 217 means 21.7°C
   int temperature : 12;
   
   //1-99 humidity in %
   byte humidity : 8;
} BL999Info;
                </pre>
            </td>
        </tr>
        <tr>
            <td>void bl999_set_rx_pin(byte pin)</td>
            <td>set up digital pin which will be used to 
            receive sensor signals. <b>Note:</b> digital pin should support ISR, 
                see datasheet for your microcontroller to find pins which supports interrupts.
                See <a href="https://www.arduino.cc/en/Reference/AttachInterrupt">Arduino ISR pins</a>
                </td>
            <td></td>
        </tr>
        <tr>
            <td>void bl999_rx_start()</td>
            <td>starts listening for messages from BL999 sensor(s)</td>
            <td></td>
        </tr>
        <tr>
            <td>void bl999_rx_stop()</td>
            <td>stops listening for the messages from sensor(s) </td>
            <td></td>
        </tr>
        <tr>
            <td>void bl999_wait_rx()</td>
            <td>blocks execution until message from sensor(s) will be received</td>
            <td></td>
        </tr>
        <tr>
            <td>boolean bl999_wait_rx_max(unsigned long milliseconds)</td>
            <td>blocks execution until message from sensor(s) will be received
            but not more than for passed amount of milliseconds. Returns <code>true</code> if message has been receieved
                <code>false</code> otherwise</td>
            <td></td>
        </tr>
        <tr>
            <td>boolean bl999_have_message()</td>
            <td>returns <code>true</code> if message from any sensor was recieved, <code>false</code> otherwise.
            <b>NOTE:</b> message will not be overridden with furrther messages
                until it will be read by the client using <code>bl999_get_message(BL999Info&amp; info)</code> method.
                <b>NOTE2:</b> this function does not take in account check sum, so it will return <code>true</code> even if message check sum is incorrect</td>
            <td></td>
        </tr>
        <tr>
            <td>boolean bl999_get_message(BL999Info&amp; info)</td>
            <td>if message was recieved (matches check sum or not)
                it will be written to the passed <code>info</code> structure.
                Returns <code>true</code> if message received and check sum matches
                returns <code>false</code> otherwise</td>
            <td></td>
        </tr>
    </tbody>
</table>



CHANGELOG.md
#####0.0.9
- Fixed keywords definitions for highlighting in the Arduino IDE

#####0.0.8
- Readme updates

#####0.0.7
- Fixed divider bit timings: 550uS+-100uS
- Fixed type conversions in Serial.prinln in examples

#####0.0.6
- Added keywords.txt to satisfy arduino library standards
- aded library.json for platformio support

#####0.0.5
- Example moved to a proper place

#####0.0.4
- Added non-blocking example

#####0.0.3
- Updated documentation

#####0.0.2
- Fixed wait functions which used infinite loop without any delays 
    which brakes some microcontrollers like esp8266

#####0.0.1
- First version:
    - Allows to configure digital pin which will be used
    - Allows to receive data from BL999


bl999.h
/*
 * bl999.h - library for Arduino to work with BL999 temperature/humidity sensor
 * See README file in this directory for more documentation
 *
 * Author: Sergey Prilukin sprilukin@gmail.com
 * $Id$
 */

#ifndef bl999_h
#define bl999_h

#include "Arduino.h"


// == PWM high and low pulse length ==

//high pulse length of the bit's divider
#define BL999_DIVIDER_PULSE_LENGTH 600

//length of the start bit low pulse
#define BL999_START_BIT_LENGTH 9000

//binary 1 low pulse length
#define BL999_BIT_1_LENGTH 3900

//binary 0 low puls length
#define BL999_BIT_0_LENGTH 1850


// == Thresholds for pulse length ==

//threshold for divider pulse
#define BL999_DIVIDER_THRESHOLD 100

//threshold for start bit pulse
#define BL999_START_BIT_THRESHOLD 1000

//threshold for 0 and 1 bit's pulse
#define BL999_REGULAR_BIT_THRESHOLD 500


// == Data array constants ==

//total amount of bits sent by BL999 sensor
#define BL999_DATA_BITS_AMOUNT 36

//36 bits are sent in packets by 4 bits (nibbles)
#define BL999_BITS_PER_PACKET 4

//totally we have 36/4 = 9 nibbles
#define BL999_DATA_ARRAY_SIZE BL999_DATA_BITS_AMOUNT / BL999_BITS_PER_PACKET

//In this struct result will be stored
typedef struct {
    byte channel : 2; // up to 3 channels
    byte powerUUID : 6; //uniq power state per current power on of the sensor
    byte battery : 1; // 0 - ok, 1- low
    int temperature : 12; // 217 = 21.7°
    byte humidity : 8; //0-100
} BL999Info;

// Cant really do this as a real C++ class, since we need to have
// an ISR
extern "C" {

// API
extern void bl999_set_rx_pin(byte pin);

extern void bl999_rx_start();

extern void bl999_rx_stop();

extern void bl999_wait_rx();

extern boolean bl999_wait_rx_max(unsigned long milliseconds);

extern byte bl999_have_message();

extern boolean bl999_get_message(BL999Info& info);

// Private

extern void _bl999_rising();

extern void _bl999_falling();

extern boolean _bl999_isCheckSumMatch();

extern byte _bl999_getSensorChannel();

extern byte _bl999_getPowerUUID();

extern byte _bl999_getPowerStatus();

extern int _bl999_getTemperature();

extern byte _bl999_getHumidity();

extern void _bl999_fillDataArray(byte bitNumber, boolean isOne);

// Matcher for divider bit
extern boolean _bl999_matchDivider(int value);

// Matcher for start bit
extern boolean _bl999_matchStartBit(int value);

// Matcher for binary 1
extern boolean _bl999_matchOneBit(int value);

// Matcher for binary 0
extern boolean _bl999_matchZeroBit(int value);

//Whether pulse length value matches specified constant with specified threshold
extern boolean _bl999_match(int value, int mathConst, int threshold);

//set state to new value but only when condition is true
extern void _bl999_setState(byte st, boolean condition);
}

#endif

bl999.cpp
/*
 * bl999.h - library for Arduino to work with BL999 temperature/humidity sensor
 * See README file in this directory for more documentation
 *
 * Author: Sergey Prilukin sprilukin@gmail.com
 * $Id$
 */

#include <Arduino.h>
#include "bl999.h"

static volatile unsigned int bl999_pwm_high_length = 0;
static volatile unsigned int bl999_pwm_low_length = 0;
static volatile unsigned int bl999_prev_time_rising = 0;
static volatile unsigned int bl999_prev_time_falling = 0;
static volatile byte bl999_state = 0;
static byte bl999_data[BL999_DATA_ARRAY_SIZE] = {0, 0, 0, 0, 0, 0, 0, 0, 0};
static byte bl999_pin = 2;
static volatile boolean bl999_active = false;
static volatile boolean bl999_message_ready = false;

extern "C" {

//=====================
//Public API
//=====================

extern void bl999_set_rx_pin(byte pin) {
    bl999_rx_stop();
    bl999_pin = pin;
}

extern void bl999_rx_start() {
    if (!bl999_active) {
        attachInterrupt(digitalPinToInterrupt(bl999_pin), _bl999_rising, RISING);
        bl999_state = 0;
        bl999_active = true;
    }
}

extern void bl999_rx_stop() {
    if (bl999_active) {
        detachInterrupt(digitalPinToInterrupt(bl999_pin));
        bl999_active = false;
        bl999_state = 0;
    }
}

extern void bl999_wait_rx() {
    while (bl999_active && !bl999_message_ready)
        ;
}

extern boolean bl999_wait_rx_max(unsigned long milliseconds) {
    unsigned long start = millis();

    while (bl999_active && !bl999_message_ready && ((millis() - start) < milliseconds))
        ;

    return bl999_message_ready;
}

extern byte bl999_have_message() {
    return bl999_message_ready;
}

extern boolean bl999_get_message(BL999Info& info) {

    // Message available?
    if (!bl999_message_ready) {
        return false;
    }

    info.channel = _bl999_getSensorChannel();
    info.powerUUID = _bl999_getPowerUUID();
    info.battery = _bl999_getPowerStatus();
    info.temperature = _bl999_getTemperature();
    info.humidity = _bl999_getHumidity();

    byte checkSumMatches = _bl999_isCheckSumMatch();

    bl999_state = 0;
    bl999_message_ready = false;

    return checkSumMatches;
}

//===============
//Private
//===============

extern void _bl999_rising() {
    attachInterrupt(digitalPinToInterrupt(bl999_pin), _bl999_falling, FALLING);

    //Do not rewrite last received but unread message
    if (!bl999_message_ready) {

        //will be used in falling interrupt
        bl999_prev_time_rising = micros();

        bl999_pwm_low_length = micros() - bl999_prev_time_falling;

        if (bl999_state % 2 == 0) {
            //Clear state to 0 since we are now in ivalid state
            _bl999_setState(0, true);
        } else {
            if (bl999_state == 1) {
                //We expect start bit to match current LOW pulse
                _bl999_setState(2, _bl999_matchStartBit(bl999_pwm_low_length));
                return;
            }

            byte newState = bl999_state + 1;
            boolean matchOne = _bl999_matchOneBit(bl999_pwm_low_length);
            boolean matchZero = _bl999_matchZeroBit(bl999_pwm_low_length);

            //We expect value bit (0 or 1) to match current LOW pulse
            _bl999_setState(newState, matchOne || matchZero);

            if (bl999_state == newState) {
                //Ok, we've read next value bit, lets calc it's number
                int bitNumber = (bl999_state - 1) / 2 - 1;

                //fill data array with this next value bit
                _bl999_fillDataArray(bitNumber, matchOne);

                if (bitNumber == BL999_DATA_BITS_AMOUNT - 1) {
                    //This was the last value bit

                    //All is done we can return the whole message to the client
                    //Do not calc check sum here.
                    //It will be calculated at message reading stage
                    //if (_bl999_isCheckSumMatch()) {
                        bl999_message_ready = true;
                    //}

                    //clear state to zero since nothing more we can do
                    _bl999_setState(0, true);
                    bl999_prev_time_rising = 0;
                    bl999_prev_time_falling = 0;
                    bl999_pwm_high_length = 0;
                    bl999_pwm_low_length = 0;
                }
            }
        }
    }
}

extern void _bl999_falling() {
    attachInterrupt(digitalPinToInterrupt(bl999_pin), _bl999_rising, RISING);

    //Do not rewrite last received but unread message
    if (!bl999_message_ready) {
        //remember current time
        //it will be used in rising function
        bl999_prev_time_falling = micros();

        //calc length of last HIGH pulse
        bl999_pwm_high_length = micros() - bl999_prev_time_rising;

        if (bl999_state % 2 == 0) {
            //We only can proceed if state is even: 0, 2, 4 - after this state
            //we always expect divider HIGH pulse to happen
            _bl999_setState(bl999_state + 1, _bl999_matchDivider(bl999_pwm_high_length));
        } else {
            //Clear state to zero since it happens that we are in invalid state
            //at this point
            _bl999_setState(0, true);
        }
    }
}

extern boolean _bl999_isCheckSumMatch() {

    //Sum first 8 nibbles
    int sum = 0;
    for (byte i = 0; i < BL999_DATA_ARRAY_SIZE - 1; i++) {
        sum += bl999_data[i];
    }

    //clear higher bits
    sum &= 15;

    //returns true if calculated check sum matches received
    return sum == bl999_data[BL999_DATA_ARRAY_SIZE - 1];
}

extern byte _bl999_getSensorChannel() {
    //Channel is bits B0, B1 in T2 nibble
    //Since we store T2 in reversed order
    //we have to get 3,4 bits ot data[1] and reverse them
    return ((bl999_data[1] & 1) << 1) | ((bl999_data[1] & 2) >> 1);
}

extern byte _bl999_getPowerUUID() {
    //we do not reverse power uuid since value is still uniq
    return (bl999_data[0] << 2) | ((bl999_data[1] & 12) >> 2);
}

extern byte _bl999_getPowerStatus() {
    //Power or battery status is C0 bit in T3 nibble
    //since in data array we wrote T3 in reversed order
    //it will be the first bit in data[2]
    return bl999_data[2] & 1;
}

extern int _bl999_getTemperature() {
    int temperature = 0;

    //Temperature is stored in T4,T5,T6 nibbles
    //lowest nibble - first
    //since we already reversed bits order in these nibbles
    //all we have to do is to reverse nibbles order
    temperature = (((int)bl999_data[5] << 8)
                   | ((int)bl999_data[4] << 4)
                   | (int)bl999_data[3]);

    if ((bl999_data[5] & 1) == 1) {
        //negative number, use two's compliment conversion
        temperature = ~temperature + 1;

        //clear higher bits and convert to negative
        temperature = -1 * (temperature & 4095);
    }

    return temperature;
}

extern byte _bl999_getHumidity() {

    //Humidity is stored in nibbles T7,T8
    //since bits in these nibbles are already in reversed order
    //we just have to get number by reversing nibbles orderßßß
    int humidity = ((int)bl999_data[7] << 4) | (int)bl999_data[6];

    //negative number, use two's compliment conversion
    humidity = ~humidity + 1;

    //humidity is stored as 100 - humidity
    return 100 - (byte)humidity;
}

extern void _bl999_fillDataArray(byte bitNumber, boolean isOne) {
    byte dataArrayIndex = bitNumber / BL999_BITS_PER_PACKET;
    byte bitInNibble = bitNumber % 4;

    if (bitInNibble == 0) {
        // if it's the first bit in nibble -
        // clear nibble since it could be filled with random data at this time
        bl999_data[dataArrayIndex] = 0;
    }

    //Write all nibbles in reversed order
    //so it will be easier to do calculations later
    bl999_data[dataArrayIndex] |= (isOne << bitInNibble);
}

// Matcher for divider bit
extern boolean _bl999_matchDivider(int value) {
    return _bl999_match(value, BL999_DIVIDER_PULSE_LENGTH, BL999_DIVIDER_THRESHOLD);
}

// Matcher for start bit
extern boolean _bl999_matchStartBit(int value) {
    return _bl999_match(value, BL999_START_BIT_LENGTH, BL999_START_BIT_THRESHOLD);
}

// Matcher for binary 1
extern boolean _bl999_matchOneBit(int value) {
    return _bl999_match(value, BL999_BIT_1_LENGTH, BL999_REGULAR_BIT_THRESHOLD);
}

// Matcher for binary 0
extern boolean _bl999_matchZeroBit(int value) {
    return _bl999_match(value, BL999_BIT_0_LENGTH, BL999_REGULAR_BIT_THRESHOLD);
}

//Whether pulse length value matches specified constant with specified threshold
extern boolean _bl999_match(int value, int mathConst, int threshold) {
    return value > mathConst - threshold && value < mathConst + threshold;
}

//set state to new value but only when condition is true
extern void _bl999_setState(byte st, boolean condition) {
    bl999_state = condition ? st : 0;
}
}

lib_bl999.cpp
/*
 * lib_bl999.cpp - library for Arduino to work with BL999 temperature/humidity sensor
 * See https://github.com/sprilukin/lib_BL999 for more documentation
 *
 * Author: Sergey Prilukin sprilukin@gmail.com
 */

#include <Arduino.h>
#include "lib_bl999.h"

static volatile unsigned long bl999_pwm_high_length = 0;
static volatile unsigned long bl999_pwm_low_length = 0;
static volatile unsigned long bl999_prev_time_rising = 0;
static volatile unsigned long bl999_prev_time_falling = 0;
static volatile byte bl999_state = 0;
static byte bl999_data[BL999_DATA_ARRAY_SIZE] = {0, 0, 0, 0, 0, 0, 0, 0, 0};
static byte bl999_pin = 2;
static volatile boolean bl999_active = false;
static volatile boolean bl999_message_ready = false;

extern "C" {

//=====================
//Public API
//=====================

extern void bl999_set_rx_pin(byte pin) {
    bl999_rx_stop();
    bl999_pin = pin;
}

extern void bl999_rx_start() {
    if (!bl999_active) {
        attachInterrupt(digitalPinToInterrupt(bl999_pin), _bl999_rising, RISING);
        bl999_state = 0;
        bl999_active = true;
    }
}

extern void bl999_rx_stop() {
    if (bl999_active) {
        detachInterrupt(digitalPinToInterrupt(bl999_pin));
        bl999_active = false;
        bl999_state = 0;
    }
}

extern void bl999_wait_rx() {
    while (bl999_active && !bl999_message_ready) {
        delay(BL999_DELAY_ON_WAIT);
    }
}

extern boolean bl999_wait_rx_max(unsigned long milliseconds) {
    unsigned long start = millis();

    while (bl999_active && !bl999_message_ready && ((millis() - start) < milliseconds)) {
        delay(BL999_DELAY_ON_WAIT);
    }

    return bl999_message_ready;
}

extern boolean bl999_have_message() {
    return bl999_message_ready;
}

extern boolean bl999_get_message(BL999Info& info) {

    // Message available?
    if (!bl999_message_ready) {
        return false;
    }

    info.channel = _bl999_getSensorChannel();
    info.powerUUID = _bl999_getPowerUUID();
    info.battery = _bl999_getPowerStatus();
    info.temperature = _bl999_getTemperature();
    info.humidity = _bl999_getHumidity();

    byte checkSumMatches = _bl999_isCheckSumMatch();

    bl999_state = 0;
    bl999_message_ready = false;

    return checkSumMatches;
}

//===============
//Private
//===============

extern void _bl999_rising() {
    attachInterrupt(digitalPinToInterrupt(bl999_pin), _bl999_falling, FALLING);

    //Do not rewrite last received but unread message
    if (!bl999_message_ready) {

        //will be used in falling interrupt
        bl999_prev_time_rising = micros();

        bl999_pwm_low_length = micros() - bl999_prev_time_falling;

        if (bl999_state % 2 == 0) {
            //Clear state to 0 since we are now in ivalid state
            _bl999_setState(0, true);
        } else {
            if (bl999_state == 1) {
                //We expect start bit to match current LOW pulse
                _bl999_setState(2, _bl999_matchStartBit(bl999_pwm_low_length));
                return;
            }

            byte newState = bl999_state + 1;
            boolean matchOne = _bl999_matchOneBit(bl999_pwm_low_length);
            boolean matchZero = _bl999_matchZeroBit(bl999_pwm_low_length);

            //We expect value bit (0 or 1) to match current LOW pulse
            _bl999_setState(newState, matchOne || matchZero);

            if (bl999_state == newState) {
                //Ok, we've read next value bit, lets calc it's number
                int bitNumber = (bl999_state - 1) / 2 - 1;

                //fill data array with this next value bit
                _bl999_fillDataArray(bitNumber, (byte)matchOne);

                if (bitNumber == BL999_DATA_BITS_AMOUNT - 1) {
                    //This was the last value bit

                    //All is done we can return the whole message to the client
                    //Do not calc check sum here.
                    //It will be calculated at message reading stage
                    //if (_bl999_isCheckSumMatch()) {
                        bl999_message_ready = true;
                    //}

                    //clear state to zero since nothing more we can do
                    _bl999_setState(0, true);
                    bl999_prev_time_rising = 0;
                    bl999_prev_time_falling = 0;
                    bl999_pwm_high_length = 0;
                    bl999_pwm_low_length = 0;
                }
            }
        }
    }
}

extern void _bl999_falling() {
    attachInterrupt(digitalPinToInterrupt(bl999_pin), _bl999_rising, RISING);

    //Do not rewrite last received but unread message
    if (!bl999_message_ready) {
        //remember current time
        //it will be used in rising function
        bl999_prev_time_falling = micros();

        //calc length of last HIGH pulse
        bl999_pwm_high_length = micros() - bl999_prev_time_rising;

        if (bl999_state % 2 == 0) {
            //We only can proceed if state is even: 0, 2, 4 - after this state
            //we always expect divider HIGH pulse to happen
            _bl999_setState(bl999_state + 1, _bl999_matchDivider(bl999_pwm_high_length));
        } else {
            //Clear state to zero since it happens that we are in invalid state
            //at this point
            _bl999_setState(0, true);
        }
    }
}

extern boolean _bl999_isCheckSumMatch() {

    //Sum first 8 nibbles
    int sum = 0;
    for (byte i = 0; i < BL999_DATA_ARRAY_SIZE - 1; i++) {
        sum += bl999_data[i];
    }

    //clear higher bits
    sum &= 15;

    //returns true if calculated check sum matches received
    return sum == bl999_data[BL999_DATA_ARRAY_SIZE - 1];
}

extern byte _bl999_getSensorChannel() {
    //Channel is bits B0, B1 in T2 nibble
    //Since we store T2 in reversed order
    //we have to get 3,4 bits ot data[1] and reverse them
    return ((bl999_data[1] & 1) << 1) | ((bl999_data[1] & 2) >> 1);
}

extern byte _bl999_getPowerUUID() {
    //we do not reverse power uuid since value is still uniq
    return (bl999_data[0] << 2) | ((bl999_data[1] & 12) >> 2);
}

extern byte _bl999_getPowerStatus() {
    //Power or battery status is C0 bit in T3 nibble
    //since in data array we wrote T3 in reversed order
    //it will be the first bit in data[2]
    return bl999_data[2] & 1;
}

extern int _bl999_getTemperature() {
    int temperature = 0;

    //Temperature is stored in T4,T5,T6 nibbles
    //lowest nibble - first
    //since we already reversed bits order in these nibbles
    //all we have to do is to reverse nibbles order
    temperature = (((int)bl999_data[5] << 8)
                   | ((int)bl999_data[4] << 4)
                   | (int)bl999_data[3]);

    if ((bl999_data[5] & 1) == 1) {
        //negative number, use two's compliment conversion
        temperature = ~temperature + 1;

        //clear higher bits and convert to negative
        temperature = -1 * (temperature & 4095);
    }

    return temperature;
}

extern byte _bl999_getHumidity() {

    //Humidity is stored in nibbles T7,T8
    //since bits in these nibbles are already in reversed order
    //we just have to get number by reversing nibbles orderßßß
    int humidity = ((int)bl999_data[7] << 4) | (int)bl999_data[6];

    //negative number, use two's compliment conversion
    humidity = ~humidity + 1;

    //humidity is stored as 100 - humidity
    return 100 - (byte)humidity;
}

extern void _bl999_fillDataArray(byte bitNumber, byte value) {
    byte dataArrayIndex = bitNumber / BL999_BITS_PER_PACKET;
    byte bitInNibble = bitNumber % 4;

    if (bitInNibble == 0) {
        // if it's the first bit in nibble -
        // clear nibble since it could be filled with random data at this time
        bl999_data[dataArrayIndex] = 0;
    }

    //Write all nibbles in reversed order
    //so it will be easier to do calculations later
    bl999_data[dataArrayIndex] |= (value << bitInNibble);
}

// Matcher for divider bit
extern boolean _bl999_matchDivider(unsigned long value) {
    return _bl999_match(value, BL999_DIVIDER_PULSE_LENGTH, BL999_DIVIDER_THRESHOLD);
}

// Matcher for start bit
extern boolean _bl999_matchStartBit(unsigned long value) {
    return _bl999_match(value, BL999_START_BIT_LENGTH, BL999_START_BIT_THRESHOLD);
}

// Matcher for binary 1
extern boolean _bl999_matchOneBit(unsigned long value) {
    return _bl999_match(value, BL999_BIT_1_LENGTH, BL999_REGULAR_BIT_THRESHOLD);
}

// Matcher for binary 0
extern boolean _bl999_matchZeroBit(unsigned long value) {
    return _bl999_match(value, BL999_BIT_0_LENGTH, BL999_REGULAR_BIT_THRESHOLD);
}

//Whether pulse length value matches specified constant with specified threshold
extern boolean _bl999_match(unsigned long value, unsigned long mathConst, unsigned int threshold) {
    return value > mathConst - threshold && value < mathConst + threshold;
}

//set state to new value but only when condition is true
extern void _bl999_setState(byte st, boolean condition) {
    bl999_state = condition ? st : 0;
}
}

lib_bl999.h
/*
 * lib_bl999.h - library for Arduino to work with BL999 temperature/humidity sensor
 * See https://github.com/sprilukin/lib_BL999 for more documentation
 *
 * Author: Sergey Prilukin sprilukin@gmail.com
 */

#ifndef bl999_h
#define bl999_h

#include "Arduino.h"


// == PWM high and low pulse length ==

//high pulse length of the bit's divider
#define BL999_DIVIDER_PULSE_LENGTH 550

//length of the start bit low pulse
#define BL999_START_BIT_LENGTH 9000

//binary 1 low pulse length
#define BL999_BIT_1_LENGTH 3900

//binary 0 low puls length
#define BL999_BIT_0_LENGTH 1850


// == Thresholds for pulse length ==

//threshold for divider pulse
#define BL999_DIVIDER_THRESHOLD 100

//threshold for start bit pulse
#define BL999_START_BIT_THRESHOLD 1000

//threshold for 0 and 1 bit's pulse
#define BL999_REGULAR_BIT_THRESHOLD 500


// == Data array constants ==

//total amount of bits sent by BL999 sensor
#define BL999_DATA_BITS_AMOUNT 36

//36 bits are sent in packets by 4 bits (nibbles)
#define BL999_BITS_PER_PACKET 4

//totally we have 36/4 = 9 nibbles
#define BL999_DATA_ARRAY_SIZE BL999_DATA_BITS_AMOUNT / BL999_BITS_PER_PACKET

// == Utility defines ==

//On some microcontrollers infinite loop cause watch dog reset,
//so introducing some small delay in loop
#define BL999_DELAY_ON_WAIT 10


//In this struct result will be stored
typedef struct {
    byte channel : 2; // up to 3 channels
    byte powerUUID : 6; //unique power state per current power on of the sensor
    byte battery : 1; // 0 - ok, 1- low
    int temperature : 12; // stored as temperature * 10: 217 = 21.7°
    byte humidity : 8; //1-99
} BL999Info;

// Cant really do this as a real C++ class, since we need to have
// an ISR
extern "C" {

// API

//set up digital pin which will be used to receive sensor signals
extern void bl999_set_rx_pin(byte pin);

//starts listening of the signals and read message from sensor(s)
extern void bl999_rx_start();

//stops listening for the signals
extern void bl999_rx_stop();

//blocks execution until message from sensor(s) will be received
extern void bl999_wait_rx();

//blocks execution until message from sensor(s) will be received
//but not more than for passed amount of milliseconds
extern boolean bl999_wait_rx_max(unsigned long milliseconds);

//returns true if whole message from sensor(s) was recieved
//NOTE: message will not be overridden with other messages
//until it will be read by the client using bl999_get_message
//NOTE2: this function does not take in account check sum
extern boolean bl999_have_message();

//if message was fully recieved (matches check sum or not)
//it will be written to info structure
//returns true if message received and check sum matches
//returns false otherwise
extern boolean bl999_get_message(BL999Info& info);

// Private

//rising ISR
extern void _bl999_rising();

//falling ISR
extern void _bl999_falling();

//calc check sum and compare with received one
extern boolean _bl999_isCheckSumMatch();

//get sensor channel from message
extern byte _bl999_getSensorChannel();

//get power UUID from message
extern byte _bl999_getPowerUUID();

//get power status form message
extern byte _bl999_getPowerStatus();

//get temperature from message
extern int _bl999_getTemperature();

//get humidity from message
extern byte _bl999_getHumidity();

//fill data array with next data bit
extern void _bl999_fillDataArray(byte bitNumber, byte value);

// Matcher for divider bit
extern boolean _bl999_matchDivider(unsigned long value);

// Matcher for start bit
extern boolean _bl999_matchStartBit(unsigned long value);

// Matcher for binary 1
extern boolean _bl999_matchOneBit(unsigned long int value);

// Matcher for binary 0
extern boolean _bl999_matchZeroBit(unsigned long int value);

//Whether pulse length value matches specified constant with specified threshold
extern boolean _bl999_match(unsigned long value, unsigned long mathConst, unsigned int threshold);

//set state to new value but only when condition is true
extern void _bl999_setState(byte st, boolean condition);
}

#endif
